internal/sops/azkv/keysource.go

```diff
@@ -0,0 +1,195 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package azkv
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"unicode/utf16"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/keyvault/keyvault"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/dimchansky/utfbom"
+)
+
+// MasterKey is an Azure Key Vault key used to encrypt and decrypt SOPS' data key.
+// The underlying authentication token can be configured using AADSettings.
+type MasterKey struct {
+	VaultURL string
+	Name     string
+	Version  string
+
+	EncryptedKey string
+
+	token *adal.ServicePrincipalToken
+}
+
+// LoadAADSettingsFromBytes attempts to load the given bytes into the given AADSettings.
+// By first decoding it if UTF-16, and then unmarshalling it into the given struct.
+// It returns an error for any failure.
+func LoadAADSettingsFromBytes(b []byte, s *AADSettings) error {
+	b, err := decode(b)
+	if err != nil {
+		return fmt.Errorf("failed to decode Azure authentication file bytes: %w", err)
+	}
+	if err = json.Unmarshal(b, s); err != nil {
+		err = fmt.Errorf("failed to unmarshal Azure authentication file: %w", err)
+	}
+	return err
+}
+
+// AADSettings contains the selection of fields from an Azure authentication file
+// required for Active Directory authentication.
+//
+// It is based on the unpublished contract in
+// https://github.com/Azure/go-autorest/blob/c7f947c0610de1bc279f76e6d453353f95cd1bfa/autorest/azure/auth/auth.go#L331-L342,
+// which seems to be due to an assumption of configuration through environment
+// variables over file based configuration.
```

Comment:
@phillebaba is there another file format more people might be familiar with that e.g. is provided by the UI (or tutorials around AKS)?

Reply:
I think we resolved this during our Slack discussion :smiley:

Reply:
https://docs.microsoft.com/en-us/azure/aks/kubernetes-service-principal?tabs=azure-cli#manually-create-a-service-principal